File: api/app/schemas/patients.py
"""환자(patients) 스키마(Pydantic) — web Zod 의 거울. 전 필드 snake_case.

⚠️ PII 경계: 응답 모델은 **`resident_no_masked`(마스킹)만** 담는다 —
`resident_no_enc`(암호문)·`resident_no_hash`(blind index)는 직렬화하지 않는다(클라 노출 금지).
생성 요청은 raw `resident_no` 를 받되 `birth_date`/`sex`는 서버가 검증된 RRN 에서 파생한다
(입력 불일치 제거). 정밀 검증(HARD/SOFT)은 services/rrn — 본 모듈은 형태·길이 + 보험유형 enum 만.
"""

from __future__ import annotations

import re
from datetime import date, datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, Field, StringConstraints, field_validator

# 앞뒤 공백 제거 문자열 — web Zod `.trim()` 정합 + 직접 API 호출 시 공백 패딩 우회 방지.
_Stripped = Annotated[str, StringConstraints(strip_whitespace=True)]

# 보험유형 — DB CHECK(0009)와 동일 집합. 한국어 표시는 UI 라벨(web).
InsuranceType = Literal["health_insurance", "medical_aid", "auto_insurance", "self_pay"]

# 혈액형(ABO+Rh) — 폐쇄 어휘. DB 컬럼은 의도적 text(0009) → **앱 계층(Pydantic Literal + web Zod)이
# 강제**(검증 서버 tier 권위, DB CHECK 미도입=마이그레이션 회피). 미상=None/생략.
BloodType = Literal["A+", "A-", "B+", "B-", "O+", "O-", "AB+", "AB-"]

# 이메일 형식(dep-free — EmailStr 미도입). web Zod refine 의 거울(빈 값=옵셔널 허용).
_EMAIL_RE = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


class PatientCreate(BaseModel):
    """환자 생성 요청(원무 직접 등록). resident_no 필수 — birth_date·sex 는 서버가 RRN 에서 파생.

    resident_no 는 하이픈 유/무 모두 허용(서비스가 정규화). 정밀 검증(HARD/SOFT)은 services/rrn 이
    수행하고 실패 시 422(invalid_rrn) — 여기선 길이 가드만(원본 값은 응답·에러에 echo 하지 않는다).
    """

    resident_no: _Stripped = Field(min_length=6, max_length=14)
    name: _Stripped = Field(min_length=1, max_length=100)
    phone: _Stripped | None = Field(default=None, max_length=20)
    address: _Stripped | None = Field(default=None, max_length=300)
    email: _Stripped | None = Field(default=None, max_length=200)
    insurance_type: InsuranceType
    insurance_no: _Stripped | None = Field(default=None, max_length=50)

    @field_validator("phone", "address", "email", "insurance_no", mode="after")
    @classmethod
    def _empty_to_none(cls, v: str | None) -> str | None:
        """빈 옵셔널을 None 으로 정규화 — 직접 API 호출의 ""(빈문자열) 저장 방지(NULL=값없음 일관).

        web 폼은 빈 옵셔널을 제거(toPatientCreatePayload)하나, 서버가 클라에 의존 않도록 방어심층.
        """
        return v or None

    @field_validator("email", mode="after")
    @classmethod
    def _check_email(cls, v: str | None) -> str | None:
        """비어있지 않은 이메일은 형식 검증(빈 값은 옵셔널 허용). 실패 → 422(검증 경계)."""
        if v and not _EMAIL_RE.match(v):
            raise ValueError("이메일 형식이 올바르지 않습니다")
        return v


class PatientSelfLinkRequest(BaseModel):
    """앱 자가가입 본인 연결 요청(Story 3.4, FR-003). 가입(세션 보유) 환자가 입력.

    `resident_no` 는 하이픈 유/무 허용(서비스가 정규화 후 `blind_index` 매칭). 정밀 검증(HARD/SOFT)
    은 services/rrn — 여기선 길이 가드만(원본은 응답·에러에 echo 안 함). `name` 은 사칭 방지 1차선
    (시뮬 시대 — 매칭 행과 성명 일치). 클라가 patient_id/auth_uid 를 제공하지 않는다(세션 uid 스코프
    — 연결 대상은 JWT 주체에서만 도출).
    """

    resident_no: _Stripped = Field(min_length=6, max_length=14)
    name: _Stripped = Field(min_length=1, max_length=100)


class PatientSelfSummary(BaseModel):
    """자가연결 확인 요약(Story 3.4) — 마스킹·식별 최소 필드. 본인 전체 데이터는 Epic 8 포털(RLS).

    임상·연락처·`auth_uid`·`_enc`/`_hash` 미포함(연결 확인에 불요 — PII 경계).
    `PatientListItem` 의 부분집합(생성시각 제외)."""

    id: UUID
    chart_no: str
    name: str
    birth_date: date
    sex: str
    resident_no_masked: str


class PatientEncounterCard(BaseModel):
    """환자 포털 '내 기록' 내원 카드(Story 8.1, FR-120) — 본인 내원 1건 요약(읽기 전용).

    세션 uid 스코프 — 서버가 auth_uid=sub 로 도출, patient_id 는 미투영(클라 미수용).
    ⚠️ PII 경계: raw RRN·연락처·암호문 미포함(비-PII 내원 메타 + 진단 마스터 부연만).
    날짜·시각은 클라가 12시간 KST 표기(UX-DR17). status=0010 enum 6값(클라가 환자 톤 라벨 매핑).
    primary_diagnosis_* = 활성 주상병 1건(0014 is_primary) + 0054 friendly_note(없으면 None).
    펼침 상세(처방·검사)는 Story 8.2 — 본 카드는 메타 + 주상병만.
    """

    id: UUID
    encounter_no: str
    status: str
    visit_type: str
    department_name: str
    doctor_name: str | None = None
    # 예약(reserved) 내원은 appointments.scheduled_start 조인 — walk_in 은 None.
    scheduled_start: datetime | None = None
    registered_at: datetime | None = None
    consult_started_at: datetime | None = None
    completed_at: datetime | None = None
    cancelled_at: datetime | None = None
    created_at: datetime
    cancel_reason: str | None = None
    primary_diagnosis_name: str | None = None
    primary_diagnosis_friendly_note: str | None = None


class PatientPrescriptionItem(BaseModel):
    """환자 포털 펼침 상세 — 처방 약 1줄(Story 8.2, FR-121). 처방상세 라인을 환자 언어로 평면화.

    ⚠️ 환자 비노출: `drug_id`·`ingredient_code`·`ordered_by`·헤더 메타 미투영(PII/내부 식별자).
    `frequency`·`usage_instruction` 은 저장값 그대로(실데이터는 한국어 '1일 3회'·'매 식후 30분').
    `dose`+`unit` 로 클라가 "1정" 조립. coverage_type=급여/비급여(non_covered 표시용·선택).
    """

    drug_name: str
    unit: str | None = None  # drugs.unit(정/캡슐/mL) — dose 와 함께 "1정" 조립
    dose: float | None = None  # 1회 투약량(numeric → float)
    frequency: str | None = None  # 저장값 그대로(한국어 산문, 코드 매핑 금지)
    usage_instruction: str | None = None  # 용법(식후/식전 — 한국어 그대로)
    duration_days: int | None = None
    coverage_type: str | None = None  # covered | non_covered(비급여 표시용)


class PatientExaminationItem(BaseModel):
    """환자 포털 펼침 상세 — 검사 1건(Story 8.2, FR-121). 검사명 + 환자용 결과 요약 + 정상/주의.

    ⚠️ 환자 비노출(구조적 차단): `findings`·`reading_conclusion`(임상 서사·감사 마스킹)·
    `fee_schedule_id`·`*_by` 미투영. 환자는 큐레이션 `patient_result_summary`(0055)·`_flag` 만 본다.
    완료 전(ordered/performed) 검사는 결과 NULL → 클라가 "결과 준비 중" 폴백.
    """

    exam_name: str  # = fee_schedules.name(행위명)
    exam_type: str  # lab | imaging
    status: str  # ordered | performed | completed
    patient_result_summary: str | None = None  # 쉬운 말 결과 요약(0055·없으면 폴백)
    patient_result_flag: str | None = None  # normal | attention | None(배지 진실 원천)
    completed_at: datetime | None = None


class PatientEncounterDetail(BaseModel):
    """환자 포털 '내 기록' 카드 펼침 상세(Story 8.2, FR-121) — 본인 내원 1건의 처방·검사.

    세션 uid 스코프 + 소유 검증(본인 내원 아니면 라우터/서비스가 404). 처방·검사 0건이면 빈 배열
    (클라가 섹션 생략 또는 안내). raw RRN·연락처·임상 서사 미포함(PII·민감 경계).
    """

    prescriptions: list[PatientPrescriptionItem]
    examinations: list[PatientExaminationItem]


class PatientPaymentCard(BaseModel):
    """환자 포털 '마이' 탭 수납 카드(Story 8.3, FR-122) — 본인 finalized 수납 1건 요약(읽기 전용).

    세션 uid 스코프 — 서버가 auth_uid=sub 로 도출, patient_id 미투영. **finalized 만 노출**(draft=
    집계중·cancelled=취소/노쇼 미발생은 제외). ⚠️ PII 경계: raw RRN·연락처·내부 `*_by` id 미포함
    (비-PII 결제 메타만). 금액=KRW 정수. 날짜·시각은 클라가 12시간 KST 표기(UX-DR17). 영수증 상세
    라우팅 키 = encounter_id(불투명 UUID). 영수증 문서(법정 서식)는 ReceiptResponse(7.5 재사용).
    """

    encounter_id: UUID
    payment_no: str | None = None
    clinic_name: str  # 요양기관명(clinic_profile.name — 단일 운영)
    department_name: str
    treatment_date: date | None = None  # 진료일(KST date)
    finalized_at: datetime | None = None  # 결제 완료 시각(정렬·표시)
    total_amount_krw: int  # 총 진료비
    paid_amount_krw: int  # 내가 낸 금액(카드 강조)
    payment_method: str | None = None  # card | cash | transfer(클라 한글화)
    status: str  # finalized 고정(라벨 일관성·향후 확장 여지)


class PatientClinicalProfileUpdate(BaseModel):
    """임상 프로필 갱신 요청(Story 3.2, FR-004). 5필드 옵셔널 — PUT 전체 교체(미전송=None=값없음).

    `blood_type` 은 폐쇄어휘(BloodType Literal) — 비정상 값은 422. 자유텍스트 4종은 max_length
    가드만. 임상필드는 암호화 대상 아님(평문 저장·반환). web Zod `clinicalProfileSchema` 의 거울.
    """

    blood_type: BloodType | None = None
    allergies: _Stripped | None = Field(default=None, max_length=1000)
    chronic_diseases: _Stripped | None = Field(default=None, max_length=1000)
    medications: _Stripped | None = Field(default=None, max_length=1000)
    notes: _Stripped | None = Field(default=None, max_length=2000)

    @field_validator("allergies", "chronic_diseases", "medications", "notes", mode="after")
    @classmethod
    def _empty_to_none(cls, v: str | None) -> str | None:
        """빈 옵셔널을 None 으로 정규화(직접 API 호출의 "" 저장 방지, NULL=값없음 일관)."""
        return v or None


class PatientResponse(BaseModel):
    """환자 응답(상세) — 마스킹 주민번호 + 임상 프로필(raw·암호문·blind index 미포함, PII 경계).

    임상 5필드(`blood_type`·`allergies`·`chronic_diseases`·`medications`·`notes`)는 Story 3.2 에서
    노출 — 컬럼은 0009 에 존재(전부 nullable). 생성 직후엔 NULL(등록은 임상필드를 받지 않음).
    """

    id: UUID
    chart_no: str
    name: str
    birth_date: date
    sex: str
    resident_no_masked: str
    phone: str | None = None
    address: str | None = None
    email: str | None = None
    insurance_type: str
    insurance_no: str | None = None
    # 임상 프로필(Story 3.2) — 자유텍스트 4종 + 폐쇄어휘 blood_type. 전부 nullable.
    blood_type: str | None = None
    allergies: str | None = None
    chronic_diseases: str | None = None
    medications: str | None = None
    notes: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class PatientRrnReveal(BaseModel):
    """주민번호 reveal 응답(Story 4.5, FR-242) — full RRN(권한 게이트 + 감사 후). 응답 바디 전용.

    ⚠️ 이 값(raw 주민번호)은 응답 바디로만 노출 — 로그·toast·에러봉투에 echo 금지(PII 경계)."""

    resident_no: str


class PatientContactReveal(BaseModel):
    """연락처 reveal 응답(Story 4.5, UX-DR22) — full phone/address/email(권한 게이트 + 감사 후)."""

    phone: str | None = None
    address: str | None = None
    email: str | None = None


class PatientListItem(BaseModel):
    """환자 목록 경량 항목 — 마스킹·식별 최소 필드(검색·확인용, 민감정보 없음)."""

    id: UUID
    chart_no: str
    name: str
    birth_date: date
    sex: str
    resident_no_masked: str
    phone: str | None = None
    is_active: bool
    created_at: datetime


class PatientPageMeta(BaseModel):
    """페이지네이션 메타(아키텍처 §Format Patterns: 목록 = {data, meta:{page,page_size,total}})."""

    page: int
    page_size: int
    total: int


class PatientPage(BaseModel):
    """환자 목록 페이지 봉투."""

    data: list[PatientListItem]
    meta: PatientPageMeta

File: web/src/lib/reception/patients.ts
import { z } from "zod";

import { apiFetch } from "@/lib/api/client";

// 환자 등록(Story 3.1) — 타입·Zod 스키마(Pydantic PatientCreate 거울)·페이로드 변환.
// 읽기/쓰기 모두 FastAPI(apiFetch). 주민번호는 raw 로 서버에 보내고(서버가 정규화·암호화), 클라는
// HARD 사전체크(3중 검증 1선)만 한다. 🚫 raw 주민번호는 로그·toast 에 남기지 않는다(PII 경계).

export type InsuranceType = "health_insurance" | "medical_aid" | "auto_insurance" | "self_pay";

export const INSURANCE_TYPES: { value: InsuranceType; label: string }[] = [
  { value: "health_insurance", label: "건강보험" },
  { value: "medical_aid", label: "의료급여" },
  { value: "auto_insurance", label: "자동차보험" },
  { value: "self_pay", label: "일반(비급여)" },
];

const INSURANCE_VALUES: string[] = INSURANCE_TYPES.map((t) => t.value);

/** FastAPI PatientResponse 거울(snake_case 유지 — camelCase 변환 금지). 민감정보는 마스킹뿐.
 *  임상 5필드(Story 3.2)는 상세 응답에 포함(목록 PatientListItem 엔 없음). */
export type Patient = {
  id: string;
  chart_no: string;
  name: string;
  birth_date: string;
  sex: string;
  resident_no_masked: string;
  phone: string | null;
  address: string | null;
  email: string | null;
  insurance_type: string;
  insurance_no: string | null;
  blood_type: string | null;
  allergies: string | null;
  chronic_diseases: string | null;
  medications: string | null;
  notes: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

/** 환자 목록 경량 항목(FastAPI PatientListItem 거울 — snake_case 유지). 마스킹·식별 단서만(검색·확인용).
 *  Patient(상세)와 별개: 임상·연락처(주소/이메일/보험) 없음. */
export type PatientListItem = {
  id: string;
  chart_no: string;
  name: string;
  birth_date: string;
  sex: string;
  resident_no_masked: string;
  phone: string | null;
  is_active: boolean;
  created_at: string;
};

/** 목록 페이지 봉투(아키텍처 §Format: {data, meta:{page,page_size,total}}). */
export type PatientPage = {
  data: PatientListItem[];
  meta: { page: number; page_size: number; total: number };
};

/** 전역 환자 검색(Story 3.5) — 이름·차트번호·연락처로 `GET /v1/patients?q=`. 상위 pageSize 건만.
 *  abort 가능(빠른 타이핑 시 이전 요청 취소 — 경쟁 결과 방지). 🚫 검색어(q=이름·연락처 PII)는
 *  로그·toast 에 남기지 않는다(서버 라우트도 동일 — 라우트는 불투명, 응답은 마스킹). */
export async function searchPatients(
  q: string,
  signal?: AbortSignal,
  pageSize = 20,
): Promise<PatientListItem[]> {
  const page = await apiFetch<PatientPage>(
    `/v1/patients?q=${encodeURIComponent(q)}&page_size=${pageSize}`,
    { signal },
  );
  return page.data;
}

// ── 혈액형(ABO+Rh) 폐쇄 어휘 — Pydantic BloodType Literal 의 거울 ───────────────
export type BloodType = "A+" | "A-" | "B+" | "B-" | "O+" | "O-" | "AB+" | "AB-";

export const BLOOD_TYPES: BloodType[] = ["A+", "A-", "B+", "B-", "O+", "O-", "AB+", "AB-"];
const BLOOD_VALUES: string[] = BLOOD_TYPES;

/** 혈액형 표시 — 빈 값/미상은 "미확인". 폐쇄어휘는 그대로 표기. */
export function bloodTypeLabel(value: string | null): string {
  return value ? value : "미확인";
}

// ── 주민번호 검증(클라 1선 — services/rrn 의 거울) ─────────────────────────────
// 성별·세기 자리 → 출생 세기. 9·0(1800년대)은 HARD 범위(1–8) 밖이라 제외(서버와 동일).
const CENTURY_BY_GENDER: Record<number, number> = {
  1: 1900, 2: 1900, 5: 1900, 6: 1900,
  3: 2000, 4: 2000, 7: 2000, 8: 2000,
};
const DAYS_IN_MONTH = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
const CHECKSUM_WEIGHTS = [2, 3, 4, 5, 6, 7, 8, 9, 2, 3, 4, 5];

/** 하이픈·공백 등 비숫자 제거(검증/전송 공용 정규화). */
export function normalizeRrn(raw: string): string {
  return (raw ?? "").replace(/\D/g, "");
}

function isValidBirthdate(d: string): boolean {
  const gender = Number(d[6]);
  const century = CENTURY_BY_GENDER[gender];
  if (century === undefined) return false;
  const year = century + Number(d.slice(0, 2));
  const month = Number(d.slice(2, 4));
  const day = Number(d.slice(4, 6));
  if (month < 1 || month > 12 || day < 1) return false;
  let max = DAYS_IN_MONTH[month - 1];
  if (month === 2 && !(year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0))) max = 28;
  return day <= max;
}

/** HARD(형식·성별/세기 자리·생년월일). 통과 시 null, 실패 시 한국어 메시지(차단). */
export function rrnHardError(raw: string): string | null {
  const d = normalizeRrn(raw);
  if (d.length !== 13) return "주민등록번호 13자리를 정확히 입력하세요";
  if (!(Number(d[6]) in CENTURY_BY_GENDER)) return "성별·세기 자리가 올바르지 않습니다";
  if (!isValidBirthdate(d)) return "생년월일이 올바르지 않습니다";
  return null;
}

/** 체크섬 일치 여부(SOFT — 불일치라도 차단하지 않음, 2020 개편 대비). 형식 오류는 HARD 가 처리. */
export function rrnChecksumOk(raw: string): boolean {
  const d = normalizeRrn(raw);
  if (d.length !== 13) return true;
  const total = CHECKSUM_WEIGHTS.reduce((s, w, i) => s + Number(d[i]) * w, 0);
  return (11 - (total % 11)) % 10 === Number(d[12]);
}

// ── 검증(Zod) — Pydantic PatientCreate 의 거울(3중 검증 클라 1선) ──────────────
// 옵셔널은 "" 허용 후 제출 시 정규화. insurance_type 은 staff.ts 관례대로 string+refine(빈 기본값 호환).
export const patientCreateSchema = z.object({
  resident_no: z
    .string()
    .trim()
    .min(1, "주민등록번호를 입력하세요")
    .superRefine((v, ctx) => {
      const err = rrnHardError(v);
      if (err) ctx.addIssue({ code: "custom", message: err });
    }),
  name: z.string().trim().min(1, "이름을 입력하세요").max(100),
  phone: z.string().trim().max(20),
  address: z.string().trim().max(300),
  // 이메일 형식(서버 Pydantic _EMAIL_RE 의 거울) — 빈 값은 옵셔널 허용, 채우면 형식 검증.
  email: z
    .string()
    .trim()
    .max(200)
    .refine((v) => !v || /^[^@\s]+@[^@\s]+\.[^@\s]+$/.test(v), "이메일 형식이 올바르지 않습니다"),
  insurance_type: z.string().refine((v) => INSURANCE_VALUES.includes(v), "보험유형을 선택하세요"),
  insurance_no: z.string().trim().max(50),
});
export type PatientCreateValues = z.infer<typeof patientCreateSchema>;

/** 생성 페이로드. 빈 옵셔널은 제거(서버 None 계약). resident_no 는 raw 전송(서버가 정규화·암호화). */
export function toPatientCreatePayload(v: PatientCreateValues): Record<string, unknown> {
  const payload: Record<string, unknown> = {
    resident_no: v.resident_no,
    name: v.name,
    insurance_type: v.insurance_type,
  };
  if (v.phone) payload.phone = v.phone;
  if (v.address) payload.address = v.address;
  if (v.email) payload.email = v.email;
  if (v.insurance_no) payload.insurance_no = v.insurance_no;
  return payload;
}

/** 보험유형 코드 → 한글 표시명. */
export function insuranceLabel(value: string): string {
  return INSURANCE_TYPES.find((t) => t.value === value)?.label ?? value;
}

/** 성별 코드 → 한글. */
export function sexLabel(sex: string): string {
  return sex === "male" ? "남" : sex === "female" ? "여" : sex;
}

// ── 임상 프로필(Story 3.2) — Zod 스키마(Pydantic PatientClinicalProfileUpdate 거울, 3중 검증 1선) ──
// blood_type 은 폐쇄어휘(빈 값=미상 허용), 자유텍스트 4종은 max_length 가드만(서버 거울).
export const clinicalProfileSchema = z.object({
  blood_type: z
    .string()
    .refine((v) => !v || BLOOD_VALUES.includes(v), "혈액형을 선택하세요"),
  allergies: z.string().trim().max(1000, "1000자 이내로 입력하세요"),
  chronic_diseases: z.string().trim().max(1000, "1000자 이내로 입력하세요"),
  medications: z.string().trim().max(1000, "1000자 이내로 입력하세요"),
  notes: z.string().trim().max(2000, "2000자 이내로 입력하세요"),
});
export type ClinicalProfileValues = z.infer<typeof clinicalProfileSchema>;

/** 환자 → 폼 기본값(현재값 프리필). null → "" (빈 입력). */
export function toClinicalProfileValues(p: Patient): ClinicalProfileValues {
  return {
    blood_type: p.blood_type ?? "",
    allergies: p.allergies ?? "",
    chronic_diseases: p.chronic_diseases ?? "",
    medications: p.medications ?? "",
    notes: p.notes ?? "",
  };
}

/** 갱신 페이로드(PUT 전체 교체). 빈 값은 null 로 전송(서버 None=값없음 계약 — 명시 삭제 지원). */
export function toClinicalProfilePayload(v: ClinicalProfileValues): Record<string, unknown> {
  return {
    blood_type: v.blood_type || null,
    allergies: v.allergies || null,
    chronic_diseases: v.chronic_diseases || null,
    medications: v.medications || null,
    notes: v.notes || null,
  };
}
